### ISSUE TITLE
[Bug] wxGUI G7.8: PseudoDC.DrawLine(): arguments did not match any overloaded

### ISSUE THREAD

**Issue by neteler** [OSGeo/grass#2048]
**Describe the bug**

Trying to open the 3D mode, the following error appears:

![wxpython_pseudo_dc_error](https://user-images.githubusercontent.com/1295172/147599218-0d516b8c-1a11-4ee3-9b72-eb6c0181af25.png)

```
Starting 3D view mode...                                                        
Traceback (most recent call last):
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/gui/wxpython/mapdisp/toolbars.py", line 244, in
OnSelectTool

self.parent.AddNviz()
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/gui/wxpython/mapdisp/frame.py", line 405, in
AddNviz

self.starting3dMode.emit(firstTime=True)
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/signal.py", line 229,
in emit

dispatcher.send(signal=self, *args, **kwargs)
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/dispatcher.py", line
344, in send

response = robustapply.robustApply(
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/robustapply.py", line
60, in robustApply

return receiver(*arguments, **named)
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/gui/wxpython/lmgr/frame.py", line 428, in
AddNvizTools

self.nviz = NvizToolWindow(
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 88, in __init__

self.AddPage(page=self._createViewPage(),
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 260, in
_createViewPage

view = ViewPositionWindow(panel, size=(175, 175),
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 5809, in
__init__

self.PostDraw()
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 5797, in
PostDraw

self.Draw(pos=(x, y), scale=True)
  File "/home/mneteler/software/grass78/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 5751, in Draw

self.pdc.DrawLine(w / 2, h / 2, x, y)
TypeError
:
PseudoDC.DrawLine(): arguments did not match any overloaded
call:
  overload 1: argument 1 has unexpected type 'float'
  overload 2: argument 1 has unexpected type 'float'
```

**To Reproduce**

1. Load raster map
2. switch to 3D Mode
3. See error

**System description (please complete the following information):**

- Operating System: Fedora 35
- bug affects both G7.8 and G8.0

```
GRASS version: 7.8.7dev                                                         
Code revision: f9f30f33f                                                        
Build date: 2021-12-14                                                          
Build platform: x86_64-pc-linux-gnu                                             
GDAL: 3.3.2                                                                     
PROJ: 8.2.0                                                                     
GEOS: 3.9.2                                                                     
SQLite: 3.36.0                                                                  
Python: 3.10.1                                                                  
wxPython: 4.0.7                                                                 
Platform: Linux-5.15.6-200.fc35.x86_64-x86_64-with-glibc2.34
```

Note: this is wxPython: 4.0.7-23 (i.e., with multiple recent patches to make the wxGUI start).

**Comment by swt2c** [comment]
This should fix it:
`self.pdc.DrawLine(w / 2, h / 2, x, y)` -> `self.pdc.DrawLine(w // 2, h // 2, x, y)`

**Comment by neteler** [comment]
Thanks - I tried but now this error appears:

```
Starting 3D view mode...                                                        
Traceback (most recent call last):
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/mapdisp/toolbars.py", line 235, in
OnSelectTool

self.parent.AddNviz()
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/mapdisp/frame.py", line 494, in
AddNviz

self.starting3dMode.emit(firstTime=True)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/signal.py", line 233,
in emit

dispatcher.send(signal=self, *args, **kwargs)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/dispatcher.py", line
344, in send

response = robustapply.robustApply(
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/robustapply.py", line
60, in robustApply

return receiver(*arguments, **named)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/lmgr/frame.py", line 488, in
AddNvizTools

self.nviz = NvizToolWindow(
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 109, in __init__

self.AddPage(page=self._createViewPage(), text=" %s " %
_("View"))
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 260, in
_createViewPage

view = ViewPositionWindow(panel, size=(175, 175),
mapwindow=self.mapWindow)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 5863, in
__init__

self.PostDraw()
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 5853, in
PostDraw

self.Draw(pos=(x, y), scale=True)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 5808, in Draw

self.pdc.DrawLine(w // 2, h // 2, x, y)
TypeError
:
PseudoDC.DrawLine(): arguments did not match any overloaded
call:
  overload 1: argument 3 has unexpected type 'float'
  overload 2: argument 1 has unexpected type 'int'
```

**Comment by swt2c** [comment]
Whoops, didn't realize x and y were floats too.  You'll probably need:
`self.pdc.DrawLine(w / 2, h / 2, x, y)` -> `self.pdc.DrawLine(w // 2, h // 2, int(x), int(y))` instead.  Also the same thing is probably going to be needed for DrawCircle on the next line.

**Comment by neteler** [comment]
Yes, that helps (I have applied both suggestions locally):

Now it stops at:

```
Starting 3D view mode...                                                        
Traceback (most recent call last):
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/mapdisp/toolbars.py", line 235, in
OnSelectTool

self.parent.AddNviz()
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/mapdisp/frame.py", line 494, in
AddNviz

self.starting3dMode.emit(firstTime=True)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/signal.py", line 233,
in emit

dispatcher.send(signal=self, *args, **kwargs)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/dispatcher.py", line
344, in send

response = robustapply.robustApply(
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/etc/python/grass/pydispatch/robustapply.py", line
60, in robustApply

return receiver(*arguments, **named)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/lmgr/frame.py", line 488, in
AddNvizTools

self.nviz = NvizToolWindow(
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 109, in __init__

self.AddPage(page=self._createViewPage(), text=" %s " %
_("View"))
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/nviz/tools.py", line 348, in
_createViewPage

self.FindWindowById(self.win["view"]["z-
exag"]["slider"]).SetValue(1)
  File "/home/mneteler/software/grass80/dist.x86_64-pc-
linux-gnu/gui/wxpython/gui_core/widgets.py", line 449, in
SetValue

super(FloatSlider, self).SetValue(value)
TypeError
:
Slider.SetValue(): argument 1 has unexpected type 'float'
```

which is here:

https://github.com/OSGeo/grass/blob/bc75077672a9153f67a71fb4263f5bb9c0de8588/gui/wxpython/gui_core/widgets.py#L449

**Comment by swt2c** [comment]
`super(FloatSlider, self).SetValue(value)` -> `super(FloatSlider, self).SetValue(int(value))`

**Comment by swt2c** [comment]
The `SetRange()` call above it might also need the same fix.

**Comment by neteler** [comment]
Thanks to PR #2050 the 3D mode works again in G8.0.

A backport of #2050 will likely fix it in G7.8.7 as well, keeping it open.

**Comment by neteler** [comment]
Problem solved in ecd13587d7ccde56b9f443ee884870076983c471
